source/AI.cpp

```diff
@@ -3313,32 +3313,103 @@ void AI::MovePlayer(Ship &ship, const PlayerInfo &player, Command &activeCommand
 		// Board the nearest disabled ship, focusing on hostiles before allies. Holding
 		// `Shift` results in boarding only player-owned escorts in need of assistance.
 		shared_ptr<const Ship> target = ship.GetTargetShip();
-		if(!target || !CanBoard(ship, *target) || (shift && !target->IsYours()))
+		if(!target || (target && activeCommands.Has(Command::WAIT)) || !CanBoard(ship, *target) || (shift && !target->IsYours()))
 		{
 			if(shift)
 				ship.SetTargetShip(shared_ptr<Ship>());
 
-			double closest = numeric_limits<double>::infinity();
 			bool foundEnemy = false;
-			bool foundAnything = false;
-			for(const shared_ptr<Ship> &other : ships)
-				if(CanBoard(ship, *other))
+			bool boardClosest = Preferences::Has("Board target");
+
+			auto strategy = [&]() -> function<double(const Ship &)>
+			{
+				Point current = ship.Position();
+				// If "board target" is set to false, value priority is used.
+				if(!boardClosest)
 				{
-					if(shift && !other->IsYours())
-						continue;
+					return [this, &ship](const Ship &other) -> double
+					{
+						// Use the exact cost if the ship was scanned, otherwise use an estimation.
+						return this->Has(ship, other.shared_from_this(), ShipEvent::SCAN_OUTFITS) ?
+							other.Cost() : (other.ChassisCost() * 2.);
+					};
+				}
+				// Default to distance-based strategy.
+				return [current](const Ship &other) -> double
+				{
+					return current.DistanceSquared(other.Position());
+				};
+			}();
+
+			auto boardable = vector<pair<const Ship *, double>>{};
+
+			auto fillBoardable = [&ship, &foundEnemy, &boardable, &strategy](const Ship &other)
+			{
+				if(CanBoard(ship, other))
+					boardable.emplace_back(make_pair(&other, strategy(other)));
```

Comment:
I think `piecewise_construct` makes more sense and is more in line with what `emplace_back` is intended for (calling a converting constructor). This line currently invokes the move constructor for std::pair.

Reply:
I never heard of that before today, and I can't find any examples with pair, and trying it randomly doesn't work either
It seems I don't even need to use make_pair?
Should this be done using std::forward?

Reply:
Also I'm a bit confused; isn't move constructor good? It isn't a copy constructor at least, right?

Reply:
It saves a function call. Using piecewise_construct you're constructing the pair object directly from the argument instead of move assigning a pair. No move > move > copy

Reply:
I wish I knew how to use it though

Reply:
Indeed we don't need it in this case. Really we only need it when the pair is of more complex types where the constructors need multiple parameters.